app/src/main/java/net/vonforst/evmap/storage/ChargeLocationsDao.kt

```diff
@@ -20,19 +20,23 @@ import net.vonforst.evmap.viewmodel.Resource
 import net.vonforst.evmap.viewmodel.Status
 import net.vonforst.evmap.viewmodel.await
 import net.vonforst.evmap.viewmodel.getClusterDistance
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import java.util.*
 import kotlin.math.sqrt
 
 @Dao
 abstract class ChargeLocationsDao {
     @Insert(onConflict = OnConflictStrategy.REPLACE)
     abstract suspend fun insert(vararg locations: ChargeLocation)
 
-    // TODO: add max age here
-    @Query("SELECT EXISTS(SELECT 1 FROM chargelocation WHERE dataSource == :dataSource AND id == :id AND isDetailed == 1)")
-    abstract suspend fun checkExistsDetailed(id: Long, dataSource: String): Boolean
+    @Query("SELECT EXISTS(SELECT 1 FROM chargelocation WHERE dataSource == :dataSource AND id == :id AND isDetailed == 1 )") // AND timeRetrieved > :after )")
+    abstract suspend fun checkExistsDetailed(id: Long, dataSource: String/*, after: Long*/): Boolean
 
     suspend fun insertOrReplaceIfNoDetailedExists(vararg locations: ChargeLocation) {
         locations.forEach {
+            // I would if (it.isDetailed || !checkExistsDetailed(it.id, it.dataSource, afterDate(it.dataSource))) {
+            // along with the above :after query, but it errors...?
```

Comment:
hmm, what kind of error? Did you put the `AND timeRetrieved > :after` within the braces of the `SELECT EXISTS( ... )`?

Reply:
I don't get it, I would set it as a static method if this were Java, afterDate(source) function...

Reply:
The Kotlin equivalent of static functions are "companion objects". But in this case you can't make the function static as it depends on fields of the `ChargeLocationRepository` object. The more consistent way would be to add a parameter `after: Long` to the `insertOrReplaceIfNoDetailedExists(locations)` function and calculate the `afterDate` from within the `ChargeLocationsRepository` class - just like you already did for the other functions.

Reply:
I tried this, now it is giving some obscure error on a try { ... area.

Reply:
I tested your current branch and it is compiling fine for me now... Maybe try to clean and then build again?

Reply:
It seems right now, and I adjusted code in response to what you suggested.